Analysis/src/Autocomplete.cpp

```diff
@@ -1055,22 +1056,56 @@ static AutocompleteEntryMap autocompleteStatement(
     for (const auto& kw : kStatementStartingKeywords)
         result.emplace(kw, AutocompleteEntry{AutocompleteEntryKind::Keyword});
 
-    for (auto it = ancestry.rbegin(); it != ancestry.rend(); ++it)
-    {
-        if (AstStatForIn* statForIn = (*it)->as<AstStatForIn>(); statForIn && !statForIn->hasEnd)
-            result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
-        else if (AstStatFor* statFor = (*it)->as<AstStatFor>(); statFor && !statFor->hasEnd)
-            result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
-        else if (AstStatIf* statIf = (*it)->as<AstStatIf>(); statIf && !statIf->hasEnd)
-            result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
-        else if (AstStatWhile* statWhile = (*it)->as<AstStatWhile>(); statWhile && !statWhile->hasEnd)
-            result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
-        else if (AstExprFunction* exprFunction = (*it)->as<AstExprFunction>(); exprFunction && !exprFunction->hasEnd)
-            result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
-        if (FFlag::LuauAutocompleteDoEnd)
+    if (FFlag::LuauClipExtraHasEndProps)
+    {
+        for (auto it = ancestry.rbegin(); it != ancestry.rend(); ++it)
+        {
+            if (AstStatForIn* statForIn = (*it)->as<AstStatForIn>(); statForIn && !statForIn->body->hasEnd)
+                result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
+            else if (AstStatFor* statFor = (*it)->as<AstStatFor>(); statFor && !statFor->body->hasEnd)
+                result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
+            else if (AstStatIf* statIf = (*it)->as<AstStatIf>())
+            {
+                bool hasEnd = statIf->thenbody->hasEnd;
+                if (statIf->elsebody)
+                {
+                    if (AstStatBlock* elseBlock = statIf->elsebody->as<AstStatBlock>())
+                        hasEnd = elseBlock->hasEnd;
+                }
+
+                if (!hasEnd)
+                    result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
+            }
+            else if (AstStatWhile* statWhile = (*it)->as<AstStatWhile>(); statWhile && !statWhile->body->hasEnd)
+                result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
+            else if (AstExprFunction* exprFunction = (*it)->as<AstExprFunction>(); exprFunction && !exprFunction->body->hasEnd)
+                result.emplace("end", AutocompleteEntry{AutocompleteEntryKind::Keyword});
```

Comment:
question: will this be deleted once `LuauAutocompleteDoEnd` is clipped? since its all covered by the AstStatBlock? (i guess apart from AstStatIf)

Reply:
I hadn't actually thought about that.  I think so in theory?

I'll give it a shot when it's time to clip the flag.